zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java

```diff
@@ -184,24 +185,34 @@ public Object getEntity(EntityType entityType, long entityId) {
     public void logTextFlowStateUpdate(@Observes(during = TransactionPhase.AFTER_SUCCESS) TextFlowTargetStateEvent event_) {
         // workaround for https://issues.jboss.org/browse/WELD-2019
         final TextFlowTargetStateEvent event = event_;
-        Long actorId = event.getActorId();
+
+        Long actorId = event.getKey().getActorId();
+
         if (actorId != null) {
             Lock lock = activityLockManager.getLock(actorId);
             lock.lock();
             try {
-                transactionUtil.run(() -> {
-                    HTextFlowTarget target =
-                            textFlowTargetDAO.findById(event.getTextFlowTargetId(),
-                                    false);
-                    HDocument document = documentDAO.getById(event.getDocumentId());
-                    ActivityType activityType =
-                            event.getNewState().isReviewed() ? ActivityType.REVIEWED_TRANSLATION
-                                    : ActivityType.UPDATE_TRANSLATION;
-
-                    logActivityAlreadyLocked(actorId,
-                            document.getProjectIteration(), target, activityType,
-                            target.getTextFlow().getWordCount().intValue());
-                });
+                HDocument document =
+                    documentDAO.getById(event.getKey().getDocumentId());
+
+                for (TextFlowTargetStateEvent.TextFlowTargetState state : event
+                    .getStates()) {
+                    transactionUtil.run(() -> {
+                        HTextFlowTarget target =
+                            textFlowTargetDAO.findById(
+                                state.getTextFlowTargetId(), false);
+                        ActivityType activityType =
+                            state.getNewState().isReviewed()
+                                ? ActivityType.REVIEWED_TRANSLATION
+                                : ActivityType.UPDATE_TRANSLATION;
+
+                        logActivityAlreadyLocked(actorId,
```

Comment:
The activity type is `ActivityType.UPDATE_TRANSLATION` which expect single translation. If we want to refactor this, then it should be another pull request since it will affect the dashboard display.